### DISCUSSION TITLE
Background process in .NET MAUI

### DISCUSSION THREAD

**Discussion by hrahmansha** [dotnet/maui#25802] [General]
I am developing a .NET MAUI application targeting **Windows** and **MacOS**. Currently, I am working on a feature to create a data collection module that continuously gathers system data for use throughout the application. This module will be managed by the main module. The main module will start/stop the data collection and send scan configuration settings.

The application is already experiencing performance issues with existing features, so I am considering running the data collection task in a separate process instead of a thread.

I found that the Shiny.NET library has background [Jobs](https://shinylib.net/client/jobs/) which does similar work, but it does not have support for Windows.

What are the potential methods for running my data collection module as a separate process that can interact with (receive commands from and send data to) my MAUI application?

**Comment by davepruitt** [comment]
I don't know about a separate "process", but I typically use BackgroundWorker whenever I need to do stuff in a separate thread.

**Reply by stephenquan** [reply]
You can create a singleton service with methods to do the start/stop gathering system data in a separate thread. You can use EventHandlers or WeakReferenceMessenger to broadcast events from your service to the rest of the application. Subscribers to the service can use a relevant UI thread Dispatcher to post updates to view models connected to your UI on the UI thread.

**Comment by DanielCauser** [comment]
This might be helpful to you.
https://shinylib.net/client/jobs/

**Reply by kcIcon** [reply]
It only supports up to .net 8 though
